### PR TITLE
Improve loading time

### PR DESCRIPTION
**Pull request by Tokazama**
This supersedes #62 with a better approach to decreasing invalidations. Instead of removing methods for converting `StaticInt` to `Int` this subtypes `Number`. This lets us keep a lot of the basic math methods at the loss of some generic index like behavior, (which we were losing anyway when removing natural conversion to `Int`).

~~This gets rid of __ALL__ invalidations on v1.7 and leaves the following on v1.8 and v1.9~~
<details>
<summary>SnoopCompile output</summary>

```julia
 inserting UnitRange{T}(start::Static.StaticNumber, stop) where T<:Real in Static at /Users/zchristensen/projects/Static-proposals/Static.jl/src/Static.jl:300 invalidated:
   mt_backedges: 1: signature Tuple{Type{UnitRange{_A}} where _A, Any, Any} triggered MethodInstance for UnitRange{T}(::Integer, ::Integer) where T<:Real (0 children)
                 2: signature Tuple{Type{UnitRange{_A}} where _A, Any, Any} triggered MethodInstance for UnitRange{T}(::Any, ::Any) where T<:Real (3 children)
   35 mt_cache

 inserting ifelse(::False, x, y) in Static at /Users/zchristensen/projects/Static-proposals/Static.jl/src/Static.jl:85 invalidated:
   mt_backedges: 1: signature Tuple{typeof(ifelse), Any, Real, Real} triggered MethodInstance for min(::T, ::T) where T<:Real (0 children)
                 2: signature Tuple{typeof(ifelse), Any, AbstractFloat, AbstractFloat} triggered MethodInstance for Base._fast(::typeof(max), ::AbstractFloat, ::AbstractFloat) (0 children)
                 3: signature Tuple{typeof(ifelse), Any, AbstractFloat, AbstractFloat} triggered MethodInstance for Base._fast(::typeof(min), ::AbstractFloat, ::AbstractFloat) (0 children)
                 4: signature Tuple{typeof(ifelse), Any, Int64, Int64} triggered MethodInstance for filter!(::Function, ::Vector{Union{Nothing, Base.UUID}}) (0 children)
                 5: signature Tuple{typeof(ifelse), Any, Int64, Int64} triggered MethodInstance for filter!(::Function, ::Vector{Base.UUID}) (0 children)
                 6: signature Tuple{typeof(ifelse), Any, Int64, Int64} triggered MethodInstance for filter!(::Function, ::Vector{Symbol}) (0 children)
                 7: signature Tuple{typeof(ifelse), Any, Int64, Int64} triggered MethodInstance for filter!(::Base.var"#37#39", ::Vector{Any}) (8 children)
   3 mt_cache
```

</details>

~~I've already made a PR where I tracked down the source of these in to poor inference when filtering when running Pkg. (https://github.com/JuliaLang/julia/pull/45452)~~

EDIT: PR accepted

Here are the load times:
* v1.7
  - pr: 0.028802 seconds (42.68 k allocations: 2.577 MiB, 19.91% compilation time)
  - master: 0.049850 seconds (89.03 k allocations: 5.158 MiB, 9.87% compilation time)
* v1.8
  - pr: 0.029359 seconds (46.13 k allocations: 2.806 MiB, 17.48% compilation time)
    - still has invalidations now fixed in master 
  - master: 0.058451 seconds (100.29 k allocations: 5.878 MiB, 8.20% compilation time)
* v1.9
  - pr: 0.033590 seconds (46.10 k allocations: 2.806 MiB)
  - master: 0.032208 seconds (89.56 k allocations: 5.784 MiB)

These are updated after [a70c8e3](https://github.com/SciML/Static.jl/pull/64/commits/a70c8e32242ce373c837fec8aebb6ca8f3e1395e).
I'm not working on a computer that's great for benchmarking and the time in seconds isn't very accurate, so it's probably more useful to compare allocations between this PR and master

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
This does seem like a pretty good solution.

**Comment by ChrisRackauckas** [comment]
Can we put downstream tests on here?

**Comment by Tokazama** [comment]
I'll see if I can preemptively reduce some of these errors in ArrayInterface

**Comment by chriselrod** [comment]
Mind bumping the minor version, to indicate this is a breaking release?

**Comment by Tokazama** [comment]
There are some things I still don't have a solution to here. For example, passing StaticInt to CartesianIndex requires a new method but that creates over 100 invalidations.

**Comment by chriselrod** [comment]
Which function is invalidated?
Is it deep enough (i.e. not something like `eachindex`) that we can just define `StaticCartesianIndex` as an alternative?

**Comment by Tokazama** [comment]
I came across the issue with [this line](https://github.com/JuliaSIMD/CloseOpenIntervals.jl/blob/1e31677e43f02dfebb75c243318f9ef9b84e84d7/test/runtests.jl#L4). It's because `AbstractCloseOpen` returns a `StaticInt` for `first`. It's probably easy to fix in that one package. I'm not sure if it will keep coming up though.

**Comment by Tokazama** [comment]
@cscherrer it might be good to get at least a preliminary look at this to make sure it gels with what you're doing

**Comment by cscherrer** [comment]
Thanks @Tokazama. Getting rid of the invalidations is great! I guess this will now have some of the same challenges as Symbolics, for example
```julia
julia> @variables i::Int
1-element Vector{Num}:
 i

julia> i isa Integer
false
```
The biggest concern here is that people get very used to writing methods like `f(i::Integer)`. With Symbolics and now Static going in this direction, there needs to be a more flexible way to write things like this.

**Comment by ChrisRackauckas** [comment]
Can you add a bunch of downstream tests? If those pass, then I think we're ready to rip off the bandaid.

**Comment by Tokazama** [comment]
> Can you add a bunch of downstream tests? If those pass, then I think we're ready to rip off the bandaid.

Right now I know that `LoopVectorization` will fail because it has structs with fields that specify an `Integer` subtype. I'm trying to do a quick sweep of that stuff and then I'll add some downstream tests

**Comment by chriselrod** [comment]
> There are some things I still don't have a solution to here. For example, passing StaticInt to CartesianIndex requires a new method but that creates over 100 invalidations.

Hmm. It is important that this
https://github.com/JuliaSIMD/LoopVectorization.jl/blob/cbb9f2324ec3cb2c4aa5523d8c4fb98147aa7855/test/offsetarrays.jl#L99
still passes its static size information to LoopVectorization.
It makes a big difference to both compile and runtimes.

**Comment by Tokazama** [comment]
> Perhaps I should add a test that CartesianIndices returns an object with static size information.

Do you mean the value returned from iterating `CartesianIndices` needs to be able to contain static info?

**Comment by chriselrod** [comment]
> > Perhaps I should add a test that CartesianIndices returns an object with static size information.
> 
> Do you mean the value returned from iterating `CartesianIndices` needs to be able to contain static info?

Yes.

**Comment by Tokazama** [comment]
`CartesianIndices` are only supposed to return `CartesianIndex`. It's just not strictly enforced. I've been playing around with an `Indices` type that acts more like Dex's index set types and carries bounds checking information. This particular issue might be a good reason to seriously consider implementing it.

File: Project.toml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 name = "Static"
 uuid = "aedffcd0-7271-4cad-89d0-dc628f76c6d3"
 authors = ["chriselrod", "ChrisRackauckas", "Tokazama"]
-version = "0.6.6"
+version = "0.7"
```
Comment (ChrisRackauckas) [inline Project.toml:4]:
```suggestion
version = "0.6.7"
```

This is just to run downstream tests. Revert after seeing the results.

**Comment by ChrisRackauckas** [comment]
Revert non-breaking version release seeing the results.

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by cscherrer** [comment]
It seems to me `Irrational` is pretty close to the way we'd want `StaticFloat64` to be set up:
```julia
julia> Irrational{:π} <: Real
true
```
I don't know of any problems this leads to (are there any?). Could this design could be adapted for `StaticFloat64`, so we could keep the current subtyping?

**Comment by Tokazama** [comment]
> It seems to me `Irrational` is pretty close to the way we'd want `StaticFloat64` to be set up:
> 
> ```julia
> julia> Irrational{:π} <: Real
> true
> ```
> 
> I don't know of any problems this leads to (are there any?). Could this design could be adapted for `StaticFloat64`, so we could keep the current subtyping?

I'm not sure I understand. Are you suggesting that we assign a symbol as the static value and then use dispatch to derive the float value, like...

```julia
struct StaticFloat{sym} <: AbstractFloat end

Base.float(::StaticFloat{:float_one}) = 1.0

```

**Comment by cscherrer** [comment]
> Are you suggesting that we assign a symbol as the static value and then use dispatch to derive the float value

No, I think the type parameter should be set up to map easily to the "dynamic" value, so in this case just a `Float64`. 

I could be missing something, but to me this seems like evidence the problem can be solved without sacrificing the type hierarchy.

**Comment by Tokazama** [comment]
> > Are you suggesting that we assign a symbol as the static value and then use dispatch to derive the float value
> 
> No, I think the type parameter should be set up to map easily to the "dynamic" value, so in this case just a `Float64`.
> 
> I could be missing something, but to me this seems like evidence the problem can be solved without sacrificing the type hierarchy.

The subtyping is what causes invalidations, not where we store the values.

**Comment by cscherrer** [comment]
> The subtyping is what causes invalidations, not where we store the values.

I see, so it's not type instability but invalidation, and irrational avoid this problem because they're in Base. Then the design is irrelevant to the problem, and the "just put it in Base" solution is (sadly) not getting much traction. I think I see now, thank you.

**Comment by Tokazama** [comment]
> > The subtyping is what causes invalidations, not where we store the values.
> 
> I see, so it's not type instability but invalidation, and irrational avoid this problem because they're in Base. Then the design is irrelevant to the problem, and the "just put it in Base" solution is (sadly) not getting much traction. I think I see now, thank you.

Exactly

**Review by chriselrod — APPROVED** [review]
Many packages probably still need updating, but the minor version bump means that's okay and it hopefully won't be too much work to see them update.

**Pull request merged**